Fixed #32676 -- Prevented migrations from rendering related field attributes when not passed during initialization.

David-Wobrock opened the pull request:
Refs https://code.djangoproject.com/ticket/32676

charettes reviewed (commented):
Merci pour le patch David :)

I think the adjustments make sense and it'd be great to add a regression test for it if possible. Let me know if you need any help with that!

David-Wobrock commented:
> 
> 
> Merci pour le patch David :)
> 
> I think the adjustments make sense and it'd be great to add a regression test for it if possible. Let me know if you need any help with that!

Actually yeah, if you could point me in the right direction that'd be great 🙏 
I'm unsure whether I should test directly in the `test_autodetector`, `field_deconstruction/tests.py (test_many_to_many_field)` or somewhere else 🤔 And mostly, I don't know what do we ensure there is no regression that wouldn't already be catched with the existing tests?
Thanks!

felixxm commented:
@David-Wobrock Thanks for this patch, unfortunately it doesn't fix the issue for me :thinking: 

For example:
```python 
class MyModel(models.Model):
    m2m_1 = models.ManyToManyField('self')
    m2m_2 = models.ManyToManyField('self', related_name='custom_name')
```
`makemigrations` creates the following initial migrations
```python
    operations = [
        migrations.CreateModel(
            name='MyModel',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('m2m_1', models.ManyToManyField(related_name='_test_one_mymodel_m2m_1_+', to='test_one.MyModel')),
                ('m2m_2', models.ManyToManyField(related_name='_test_one_mymodel_m2m_2_+', to='test_one.MyModel')),
            ],
        ),
    ]
```
running `makemigrations` with the `main` branch creates:
```python
    operations = [
        migrations.AlterField(
            model_name='mymodel',
            name='m2m_1',
            field=models.ManyToManyField(related_name='m2m_1_rel_+', to='test_one.MyModel'),
        ),
        migrations.AlterField(
            model_name='mymodel',
            name='m2m_2',
            field=models.ManyToManyField(related_name='m2m_2_rel_+', to='test_one.MyModel'),
        ),
    ]
```
and with this patch:
```python
    operations = [
        migrations.AlterField(
            model_name='mymodel',
            name='m2m_1',
            field=models.ManyToManyField(to='test_one.MyModel'),
        ),
        migrations.AlterField(
            model_name='mymodel',
            name='m2m_2',
            field=models.ManyToManyField(related_name='custom_name', to='test_one.MyModel'),
        ),
    ]
```

David-Wobrock commented:
Hi @felixxm 

Thanks for the comment. I'm wondering if I understood correctly what the ideal fix would be.
My understand of [Simon's comment](https://code.djangoproject.com/ticket/32676#comment:3) is that these attributes should not appear in the migration operation. In other words, we should not show implicit values in the migrations, but only the attributes that are explicit in the fields of the Model class. It's very possible that this is not the expected behaviour and I misunderstood the point :)

However, if that is what we want to do, I expect that every Django project using these fields will have to generate migrations removing those attributes 🤕 even if the migrations are no-ops, it's a bit annoying I guess.

I pushed some more autodetector tests, that show the new behaviour I think.

charettes commented:
> Thanks for the comment. I'm wondering if I understood correctly what the ideal fix would be.
My understand of Simon's comment is that these attributes should not appear in the migration operation. In other words, we should not show implicit values in the migrations, but only the attributes that are explicit in the fields of the Model class. It's very possible that this is not the expected behaviour and I misunderstood the point :)

You're correct @David-Wobrock. The idea is basically that all attributes that get assigned at field contributing/model rendering time for internal reasons should not be present in field deconstruction.

It's definitely an internal implementation detail here.

https://github.com/django/django/blob/7e51893911237dfca9294e3ca12163ff813fb656/django/db/models/fields/related.py#L1601-L1621

In the end if these was a proper API to hide relationships that didn't involve (ab)using `related_name` these internal hacks wouldn't be required.

@felixxm I think that having noop migration operations generated in the case is ok as it seems a good way in the direction of hiding implementation details of `ManyToManyField`.

felixxm commented on django/db/models/fields/related.py:
```diff
@@ -89,6 +89,16 @@ class RelatedField(FieldCacheMixin, Field):
     many_to_many = False
     many_to_one = False
 
+    def __init__(self,
+                 related_name=None,
+                 related_query_name=None,
+                 limit_choices_to=None,
+                 **kwargs):
```
Use hanging indentation.

felixxm replied:
I meant all changed lines not only this particular place :wink:

David-Wobrock replied:
Updated again ✅  Did I miss some others?

felixxm reviewed (commented):
@David-Wobrock Thanks :+1: I think it's not draft anymore. I can make a final review.

> @felixxm I think that having noop migration operations generated in the case is ok as it seems a good way in the direction of hiding implementation details of `ManyToManyField`.

Agreed, it's much clearer without them :+1: We should update the [backward incompatibility note](https://docs.djangoproject.com/en/dev/releases/4.0/#miscellaneous) about noop operations. I would promote it to a separate section under _"Backwards incompatible changes in 4.0"_.

felixxm commented on tests/schema/fields.py:
```diff
@@ -34,7 +34,11 @@ def __init__(self, to, db_constraint=True, swappable=True, related_name=None, re
+            related_name=related_name,
+            related_query_name=related_query_name,
+            limit_choices_to=limit_choices_to,
+            **kwargs)
```
```suggestion
            **kwargs,
        )
```

felixxm commented on django/db/models/fields/related.py:
```diff
@@ -1174,7 +1195,11 @@ def __init__(self, to, related_name=None, related_query_name=None,
+            related_name=related_name,
+            related_query_name=related_query_name,
+            limit_choices_to=limit_choices_to,
+            **kwargs)
```
```suggestion
            **kwargs,
        )
```

felixxm commented on django/db/models/fields/related.py:
```diff
@@ -471,7 +483,13 @@ def __init__(self, to, on_delete, from_fields, to_fields, rel=None, related_name
+            related_name=related_name,
+            related_query_name=related_query_name,
+            limit_choices_to=limit_choices_to,
+            **kwargs
```
```suggestion
            **kwargs,
```

felixxm commented on django/db/models/fields/related.py:
```diff
@@ -89,6 +89,18 @@ class RelatedField(FieldCacheMixin, Field):
+        related_name=None,
+        related_query_name=None,
+        limit_choices_to=None,
+        **kwargs
```
```suggestion
        **kwargs,
```

David-Wobrock replied:
Duly noted, thanks 👌 

I wasn't aware Django is pushing in this direction for indentation style (related to `black`'s style I guess) - I always thought making `flake8` happy was enough.

felixxm replied:
It's not really related with `black`. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

David-Wobrock commented:
Awesome, thanks a lot for the discussion 👍 

Push some edits and create a specific section in the release note.
I hope the tests are fine as they are, and feel free to rephrase the documentation.

Marking the PR as ready for some review :)

felixxm commented:
@David-Wobrock Thanks :+1: Unfortunately all new tests pass without this patch :disappointed:

David-Wobrock commented:
> 
> 
> @David-Wobrock Thanks 👍 Unfortunately all new tests pass without this patch 😞

Indeed, `ModelState`s alone don't render the model classes (as they don't even exist in these tests).
I removed the old tests and added a new test specifically for the deconstruction of `ManyToManyField`.
I hope that will do the trick :)

felixxm commented:
@David-Wobrock Many thanks :rocket: 

I pushed small edits.

felixxm commented:
buildbot, test on oracle.
